Consolidate versioning for GPU hardware

whchung opened the pull request:
Different methods are used in GPU common runtime, StreamExecutor, and XLA to
track GPU hardware versions. To support upcoming ROCm SteramExecutor and AMDGPU
XLA, it's now consolidated to DeviceVersion in StreamExecutor in this commit.

whchung committed:
Consolidate versioning for GPU hardware

Different methods are used in GPU common runtime, StreamExecutor, and XLA to
track GPU hardware versions. To support upcoming ROCm SteramExecutor and AMDGPU
XLA, it's now consolidated to DeviceVersion in StreamExecutor in this commit.

whchung commented:
@jlebar This is the PR created to consolidate GPU hardware versioning after our discussion in #20757. Now that particular PR could be greatly simplified. However this particular PR affects quite a few places in TensorFlow and XLA.

jlebar commented on tensorflow/stream_executor/device_description.h:
```diff
@@ -156,8 +200,15 @@ class DeviceDescription {
+  // Returns the AMDGCN ISA version if we're running on the ROCm platform.
+  // Minor version would always be zero.
+  // If an AMDGCN ISA version is not available, the major version will be
+  // zero, and the return value of DeviceVersion::is_valid() will be false.
```
* s/would always be/is always/

* s/the return value of DeviceDescription::is_valid() will be false/DeviceDescription::is_valid() will return false/

But even better, perhaps rewrite this comment so that it's not redundant with the comment on `DeviceVersion`?

> Gets the "hardware version" of this device, where we know how.  If we don't know how, the returned DeviceVersion's is_valid() returns false.

whchung commented on tensorflow/stream_executor/stream_executor_pimpl.cc:
```diff
@@ -333,10 +335,11 @@ bool StreamExecutor::GetConvolveBackwardFilterAlgorithms(
+  DeviceVersion device_version =
+      GetDeviceDescription().device_hardware_version();
   return dnn_support->GetConvolveBackwardFilterAlgorithms(
-      with_winograd_nonfused, cc_major, cc_minor, out_algorithms);
```
As it touches `dnn.h` and I'd like to save it for another patch.

jlebar replied:
sgtm

whchung commented:
@jlebar Updated the PR per review comments. Please be cautious reviewing the static member function `DeviceVersion::Parse` as it would be used by bazel-generated header file `cuda_config.h` and I'm not very sure making its return type be `StatusOr<DeviceVersion>` a good idea.

jlebar commented on tensorflow/compiler/xla/service/gpu/llvm_gpu_backend/gpu_backend_lib.cc:
```diff
@@ -86,25 +87,25 @@ static string GetLibdeviceFilename(const string& libdevice_dir_path,
-                 << ", " << compute_capability.second << ") ."
+    LOG(WARNING) << "Unknown compute capability (" << compute_capability
+                 << ") ."
                  << "Defaulting to libdevice for compute_" << libdevice_version;
```
Nit: Please reflow string.

jlebar commented on tensorflow/compiler/xla/service/gpu/llvm_gpu_backend/gpu_backend_lib.cc:
```diff
@@ -113,28 +114,28 @@ static string GetLibdeviceFilename(const string& libdevice_dir_path,
-                 << ", " << compute_capability.second << ") ."
+    LOG(WARNING) << "Unknown compute capability (" << compute_capability
+                 << ") ."
                  << "Defaulting to telling LLVM that we're compiling for sm_"
```
Please reflow string.

jlebar commented on tensorflow/contrib/fused_conv/kernels/fused_conv2d_bias_activation_op.cc:
```diff
@@ -322,11 +322,12 @@ void LaunchFusedConv2DBiasActivationOp<GPUDevice, T, BiasType, ScaleType>::
-        ctx, ((cc_major == 6 && cc_minor >= 1) || cc_major > 6),
+        ctx,
+        ((device_version.major_part == 6 && device_version.minor_part >= 1) ||
+         device_version.major_part > 6),
```
`device_version >= DeviceVersion(6, 1)`?

jlebar commented on tensorflow/core/common_runtime/gpu/gpu_device.cc:
```diff
@@ -1394,16 +1361,11 @@ Status BaseGPUDeviceFactory::GetValidDeviceIds(
-              << " minor: " << cc_minor
+              << "\nname: " << description.name()
+              << " major: " << device_version.major_part
+              << " minor: " << device_version.minor_part
```
Might as well use DeviceVersion::operator<< here.

jlebar commented on third_party/gpus/cuda_configure.bzl:
```diff
@@ -1015,7 +1015,7 @@ def _create_dummy_repository(repository_ctx):
             "%{cudnn_version}": _DEFAULT_CUDNN_VERSION,
             "%{cuda_compute_capabilities}": ",".join([
-                "CudaVersion(\"%s\")" % c
+                "DeviceVersion::Parse(\"%s\")" % c
```
We can restore the old behavior by making this `DeviceVersion::Parse(%s).ValueOrDie()`.  I admit I don't understand what this is doing or how to test it, though.  We'll need someone from the build team to look at this.

\summon @gunan

jlebar commented on tensorflow/core/common_runtime/gpu/gpu_device.cc:
```diff
@@ -1316,7 +1283,7 @@ std::vector<CudaVersion> GetSupportedCudaComputeCapabilities() {
   auto extra_capabilities = str_util::Split(extra_cuda_caps, ',');
   for (const auto& capability : extra_capabilities) {
-    cuda_caps.push_back(CudaVersion(capability));
+    cuda_caps.push_back(DeviceVersion::Parse(capability));
```
This shouldn't compile?  We need to "unwrap" the StatusOr one way or another.

jlebar commented on tensorflow/stream_executor/device_description.h:
```diff
@@ -20,18 +20,72 @@ limitations under the License.
+    port::StringPiece minor_str = version_name.substr(dot_pos + 1);
+    CHECK(port::safe_strto32(minor_str.data(), &minor_part))
+        << "Illegal version name: [" << version_name << "]";
+    return DeviceVersion(major_part, minor_part);
```
The idea was to return an error rather than CHECK (which is a fatal runtime assertion).

You'd probably want to return FailedPreconditionError, see stream_executor/lib/status.h.

jlebar commented on tensorflow/stream_executor/device_description.h:
```diff
@@ -154,10 +208,10 @@ class DeviceDescription {
-  // If a CUDA compute capability is not available, the major version will be
-  // zero, and the return value will be false.
-  bool cuda_compute_capability(int *major, int *minor) const;
+  // Returns the hardware version of the device.
```
At this point we should just remove the comment.  That's fine, it's obvious what it does.

whchung commented:
@jlebar revised the PR again per comments. I'm still somewhat uncomfortable with changes made in `cuda_configure.bzl` because for obvious reason I can't really test it.

jlebar commented on tensorflow/core/common_runtime/gpu/gpu_device.cc:
```diff
@@ -1316,7 +1283,9 @@ std::vector<CudaVersion> GetSupportedCudaComputeCapabilities() {
-    cuda_caps.push_back(CudaVersion(capability));
+    DeviceVersion device_version =
+        DeviceVersion::Parse(capability).ValueOrDie();
+    cuda_caps.push_back(device_version);
```
Nit, recommend inlining the `device_version` variable:

`cuda_caps.push_back(DeviceVersion::Parse(capability).ValueOrDie())`

jlebar commented on tensorflow/stream_executor/cuda/cuda_blas.cc:
```diff
@@ -1861,12 +1861,11 @@ bool CUDABlas::DoBlasGemm(
-  stream->parent()->GetDeviceDescription().cuda_compute_capability(&cc_major,
-                                                                   &cc_minor);
+  DeviceVersion device_version;
+  stream->parent()->GetDeviceDescription().device_hardware_version();
```
This does not look correct?  You mean to assign to the device_version variable, right?

jlebar commented on tensorflow/stream_executor/cuda/cuda_dnn.cc:
```diff
@@ -2984,7 +2984,7 @@ bool CudnnSupport::DoFusedConvolve(
     dnn::ProfileResult* output_profile_result) {
   int cc_major, cc_minor;
-  stream->parent()->GetDeviceDescription().cuda_compute_capability(&cc_major,
+  stream->parent()->GetDeviceDescription().device_hardware_version(&cc_major,
```
This should be a build error?

jlebar reviewed (changes requested):
> I'm still somewhat uncomfortable with changes made in cuda_configure.bzl because for obvious reason I can't really test it.

I also don't know how to test it, but maybe @yifeif can help.

whchung commented:
@jlebar Thanks. Updated the PR once again per review comments.

jlebar commented:
LGTM, but I'd like @yifeif or @gunan to sign off on the build system changes, so not approving.

Also given that we had build errors it would be nice if the tests would complete.  I don't know how to kick off another set of tests, though.  Or maybe that happens automatically?

tensorflowbutler commented:
Nagging Assignee @qlzh727: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

gunan reviewed (dismissed):
As long as it passes presubmits, I am ok with the changes to bzl files.
We are deprecating cmake with this release anyway, so I am willing to just check tests for that too.

qlzh727 commented:
Please resolve the merge conflict before we kick off the tests and merge process.

tensorflowbutler commented:
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

whchung commented:
@jlebar / @gunan after some unexpected hiatus I'm back working on this PR. I've rebased and fixed all known failures on CUDA path. I'm wondering could we give this PR another test on your end?

jlebar commented:
>  jlebar added the kokoro:force-run  label just now

No idea if this does what I was hoping, but it's worth a try.  :)

whchung commented:
@jlebar I just fixed errors in some test targets. Notice there is one failure which i don't think has anything to do with this PR:

```
//tensorflow/python/estimator:keras_test
```

also i noticed the syntax asked by `clang-format` is a bit different on CI now. could you let me know the latest syntax definition is?

jlebar commented:
> also i noticed the syntax asked by clang-format is a bit different on CI now. could you let me know the latest syntax definition is?

/me summons @yifeif 

> could you help re-run the tests?

OK.  @yifeif is there any way to make this process of rerunning the tests less work for us?  Like, is it possible to give whchung the ability to kick tests himself?

whchung commented:
@jlebar / @yifeif I fixed the errors found in the last round of tests. Wondering is it possible for me to apply `kokoro:force-run` label?

Also I'm looking forward to hearing from @yifeif wrt #20277.

jlebar commented:
> Wondering is it possible for me to apply kokoro:force-run label?

Yeah, I know this is not a good experience for you as our partner.  :(  We're currently discussing internally whether and how it might be technically possible to allow this, within our security constraints.  I'll let you know.

yifeif commented:
Sorry for the delay @whchung, I was on vacation last week. I replied in #20277. We are using clang-format  version 3.9 with -style=google in CI. The failure in `//tensorflow/python/estimator:keras_test` should be irrelevant.

tensorflowbutler commented:
Nagging Assignee @qlzh727: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

tensorflowbutler commented:
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

tensorflowbutler commented:
It has been 43 days that this pull-request has stalled. Please create a new pull-request with the requested changes.
